csrc/scheduler/matmul.cpp

```diff
@@ -757,25 +901,45 @@ void scheduleMatmul(Fusion* fusion, const MatmulParams& params) {
       {acr, bcr, ab, bb},
       {ParallelType::TIDy, ParallelType::TIDz});
 
-  scheduler_utils::BoundedDirectionalTransformPropagator::forward(
-      mma_result,
-      -1,
-      {d},
-      scheduler_utils::BoundedDirectionalTransformPropagator::Options()
-          .propagateParallelType()
-          .propagateToBoundary());
-
-  d->axis(-1)->parallelize(ParallelType::Vectorize);
-
+  if (params.has_smem_epilogue) {
+    smem_epilogue->setMemoryType(MemoryType::Shared);
+    swizzleSharedMemory(smem_epilogue, params);
+    scheduler_utils::BoundedDirectionalTransformPropagator::forward(
+        mma_result,
+        -1,
+        {smem_epilogue},
+        scheduler_utils::BoundedDirectionalTransformPropagator::Options()
+            .propagateParallelType()
+            .propagateToBoundary());
+    smem_epilogue->axis(-1)->parallelize(ParallelType::Vectorize);
+
+    // Schedule output tensor differently for better global memory access
+    // pattern.
+    scheduleOutputTensor(mma_result, d, gemm_tile);
+    d->axis(-1)->parallelize(ParallelType::Vectorize);
+
+    // Propagate output tensor transformations back to smem_epilogue
+    scheduler_utils::BoundedDirectionalTransformPropagator::backward(
+        d, -1, {smem_epilogue});
+  } else {
+    scheduler_utils::BoundedDirectionalTransformPropagator::forward(
+        mma_result,
+        -1,
+        {d},
+        scheduler_utils::BoundedDirectionalTransformPropagator::Options()
+            .propagateParallelType()
+            .propagateToBoundary());
+    d->axis(-1)->parallelize(ParallelType::Vectorize);
+  }
   // propagate output transformations to all inputs that are part of epilogue
   //  operations, input tvs with non-core roles
   //  core roles: essential for matmul, for example mma inputs' producers
   if (has_non_mma_input_tvs) {
     scheduleFusionInputsForEpilogue(roles_map);
   }
 
-  // auto inline for all tensors except register tensors and output tensor
-  inlineMost(ir_utils::allTvsExcept(fusion, {acr, bcr, ab, bb, d}));
+  // auto inline for all tensors except register tensors
+  inlineMost(ir_utils::allTvsExcept(fusion, {acr, bcr, ab, bb}));
```

Comment:
Should this be a conditional modifications, based on `params.has_smem_epilogue`?

Something like this:
```
  // auto inline for all tensors except register tensors
  std::vector<TensorView*> inlining_exceptions = {acr, bcr, ab, bb};
  if (!params.has_smem_epilogue) {
    // with no smemu epilogue, we should also exclude output tensor
    inlining_exceptions.push_back(d);
  }
  inlineMost(ir_utils::allTvsExcept(fusion, inlining_exceptions));
```

Reply:
I just noticed the special treatment of the output tensor is not needed. I'll reverse this change and leave it to another PR.